### ISSUE TITLE
[QUESTION] GPT Pretraining Objective

### ISSUE THREAD

**Issue by anon-nlp-dev** [NVIDIA/Megatron-LM#434]
Hi,
I am quite new to the language modeling in general, and while going through your code to see the inner working of the GPT model, I came across the following two lines: https://github.com/NVIDIA/Megatron-LM/blob/65da5be50cdaad2205e0bc96295bcaaf2c2c2a34/pretrain_gpt.py#L53-L54

Since GPT is an auto-regressive model, I was expecting the `labels` to only contain the next token, however it seems that the `labels` share entries with the input `tokens`. Can you please explain why is it done this way?

My assumption is that the next token is generated iteratively by increasing the number of input tokens and masking the rest. Afterwards, we take all the predicted tokens and align them to the labels and compute the loss. Is that the right assumption?

**Comment by janEbert** [comment]
You're right that Transformers are auto-regressive, however, that is only if you want to generate _new_ tokens. In general, Transformers will generate one output token for each input token. For decoder-only models (like GPT), during the generation for each output token, only the input tokens "before" that output token are seen due to the causal attention mask. More specifically, to generate output at position `t`, the model has access to all inputs at positions `<= t`.
Maybe the following example also helps to visualize what `tokens` and `labels` look like in practice. Here, I just made each word one token and ignore whitespace for the sake of the example. In practice, this is not how it is done, but that's another story:
```
Tokens (input): ['<bos>', 'I',    'have', 'a',   'dog', '.']
Labels:         ['I',     'have', 'a',    'dog', '.',   '<eos>']
```

As you can see, the model is expected to generate the token at the next position for each input position. We additionally have to supply a `<bos>` (beginning of sentence) token so the model has an "empty" input to generate from. After all, we would like the model to be able to generate from scratch, with no information at all. That's what the `<bos>` token enables. When the input is finished, we would like the model to also predict this via the `<eos>` (end of sentence) token. That's how we enable the model to decide to stop generation at some point.

While the `<bos>` and `<eos>` tokens are named according to "sentences", this is not how they are used in the Megatron codebase; here, they simply distinguish the beginning or end of the text the model receives.

So now why do we care about all the labels? Because it's super efficient. We generate all training signals for the whole input text with one forward pass of the model. Due to the causal masking of the Transformer decoder, the model trains to predict `'I'` from `['<bos>']`, it learns to predict `'have'` from `['<bos>', 'I']`, it learns to predict `'a'` from `['<bos>', 'I', 'have']`, and so on. All with a single forward pass over all the input tokens.

**Comment by anon-nlp-dev** [comment]
@janEbert Thank you for the detailed explanation, I appreciate you taking time to explain this. Knowing how causal masking helps in training by generating output for each of the input position in a single forward pass, helps a lot in understanding the inner working of a decoder :)
